File: static/scripts/TippersRequests.js

var WEIGHT_SENSOR_TYPE = 6
var DIST_SENSOR_TYPE = 7

//these later should be based on tippershost
var BASE_INFRASTRUCTURE_URL = "http://sensoria.ics.uci.edu:8059/infrastructure/get?";
var BASE_SENSOR_URL = "http://sensoria.ics.uci.edu:8059/sensor/get?";
var BASE_OBS_URL = "http://sensoria.ics.uci.edu:8059/observation/get?";

var TIPPERS_FORMAT = 'YYYY-MM-DD HH:mm:ss';


//call these functions like this: get_sensors(6).then(function(data){ });


/*
	@return an infrastructure json object
	http://sensoria.ics.uci.edu:8059/sensor/get?sensor_type_id=6
*/
function get_infrastructure() {
	return $.getJSON(BASE_INFRASTRUCTURE_URL).then(function(data) {
		// console.log("asdf");
		throw "asdf";
		return data;
	})
	.catch(function() {
		return [
			{
				"name":"DBH",
				"id":193813,
				"type":"building",
				"region":
					{
						"geometry":[],
						"id":412,
						"floor":0
					},
				"capacity":0
			},
			{
				"name":"ISG",
				"id":193821,
				"type":"region",
				"region":
					{
						"geometry":[],
						"id":420,
						"floor":2
					},
				"capacity":0},
			{
				"name":
				"2-floor",
				"id":193838,
				"type":"floor",
				"region":
					{
						"geometry":[],
						"id":440,
						"floor":0
					},
				"capacity":0
			},
			{"area":"ISG","name":"2065","id":28,"type":"conference_room","region":{"geometry":[{"x":55,"y":8,"z":2},{"x":60,"y":19,"z":2}],"id":28,"floor":2},"building":"DBH","capacity":20}
		]
	});
}

/*
@param sensor_type 6 for weight, 7 for distance
@return a sensor json object
http://sensoria.ics.uci.edu:8059/sensor/get?sensor_type_id=6
*/
function get_sensors(sensor_type) {
	return $.getJSON( BASE_SENSOR_URL + "sensor_type_id=" + sensor_type).then(function( data ) {
		return data;
	})
	.catch(function() {
		return [
			{
				"id": "ZBin1",
				"name": "RZotbins",
				"x": 10,
				"y": 20,
				"z": 2
			},
			{
				"id": "ZBin2",
				"name": "RZotbins",
				"x": 50,
				"y": 40,
				"z": 2
			}
		];
	});
}

//@param sensor_id the id of the sensor e.g. ZBin1
//@param type the type of the sensor (WEIGHT_OBS_TYPE = 2 DIST_OBS_TYPE = 3)
//@param start_time the timestamp in which to start getting observations
//@param end_time the timestamp in which to stop getting observations
//@return an observation json object, sorted	
//http://sensoria.ics.uci.edu:8059/observation/get?sensor_id=ZBin1&type=2&start_timestamp=2018-07-01+00%3A00%3A00&end_timestamp=2018-07-25+00%3A00%3A00
//default timestamps are within the past 24 hours
function get_observation(sensor_id, start_time=null, end_time=null){
	if(start_time === null || end_time === null){
		start_time = moment().subtract(1, 'days').format(TIPPERS_FORMAT);
		end_time = moment().format(TIPPERS_FORMAT);
	}
	return $.getJSON(BASE_OBS_URL + "sensor_id=" + sensor_id +
									"&start_timestamp=" + encodeURIComponent(start_time) +
									"&end_timestamp=" + encodeURIComponent(end_time)).then(function( data ) {
		return data;
	})
	.catch(function() {
		return [
			{
				"sensor_id": sensor_id,
				"payload": {
								"weight": 123
							},
				"request_ids": [],
				"id": 123,
				"timestamp": start_time
			},
			{
				"sensor_id": sensor_id,
				"payload": {
								"weight": 127
							},
				"request_ids": [],
				"id": 123,
				"timestamp": end_time - start_time
			}
		];
	});
}

File: TippersRequests.py
import time
from datetime import *
from utils import *
import urllib.parse
import json
import requests
import math

from flask import Blueprint, session


tippers_requests = Blueprint('tippers_requests', __name__, template_folder='templates')

#tippers api urls
BASE_ROOMS_URL = "http://sensoria.ics.uci.edu:8059/infrastructure/get?"
BASE_LOCATIONS_URL = "http://sensoria.ics.uci.edu:8059/semanticobservation/presence/getLast?"
BASE_SENSOR_URL = "http://sensoria.ics.uci.edu:8059/sensor/get?"
BASE_OBS_URL = "http://sensoria.ics.uci.edu:8059/observation/get?"

BASE_ROOMS_URL = "/infrastructure/get?"
BASE_LOCATIONS_URL = "/semanticobservation/presence/getLast?"
BASE_SENSOR_URL = "/sensor/get?"
BASE_OBS_URL = "/observation/get?"

#locations constants
LOCATION_TYPE = 4
SERVICE_ID =  7

#tippers globals
WEIGHT_OBS_TYPE = 2
DIST_OBS_TYPE = 3

WEIGHT_SENSOR_TYPE = 6
DIST_SENSOR_TYPE = 7

BIN_HEIGHT = 80

# @return all tippers rooms as a json object
# http://sensoria.ics.uci.edu:8059/infrastructure/get?
def get_tippers_rooms():
	print(session)
	return eval(requests.get(session["tippershost"] + BASE_ROOMS_URL).text)
	
# @param the floor of the rooms
# @return all tippers rooms in the given floor as a json object
# http://sensoria.ics.uci.edu:8059/infrastructure/get?
def get_tippers_rooms_by_floor(floor):
	return list(filter(lambda x: x["region"]["floor"] == floor, get_tippers_rooms()))

# @param the room to get the info for
# @return given room json object
# http://sensoria.ics.uci.edu:8059/infrastructure/get?
def get_tippers_room(room):
	parameters = urllib.parse.urlencode([("name", room)])
	return eval(requests.get("http://sensoria.ics.uci.edu:8059"  + BASE_ROOMS_URL + parameters).text)
	
	
# @param subject_id the email of the user to get the locaiton of
# @return http://sensoria.ics.uci.edu:8059/semanticobservation/presence/getLast?type=4&subject_id=dhrubajg@uci.edu&requestor_id=primal@uci.edu&service_id=7
def get_location():
	parameters = urllib.parse.urlencode([("type", LOCATION_TYPE), ("subject_id", "caoj11@uci.edu"), ("start_timestamp", "2018-07-16 00:00:00"), 
										("end_timestamp", "2018-07-16 23:00:00"), ("requestor_id", "caoj11@uci.edu"), ("service_id", SERVICE_ID)])
	try:
		return eval(requests.get("http://sensoria.ics.uci.edu:8059" + BASE_LOCATIONS_URL + parameters).text)[-1]
	# if cant find the users location
	except:
		return {"payload": {"location": 2065}}
	
	
# @param sensor_type_id WEIGHT_SENSOR_TYPE = 6, DIST_SENSOR_TYPE = 7
# @return sensors json object
# http://sensoria.ics.uci.edu:8059/sensor/get?sensor_type_id=7
def get_sensors(sensor_type_id):
	parameters = urllib.parse.urlencode([("sensor_type_id", sensor_type_id)])
	return eval(requests.get(session["tippershost"] + BASE_SENSOR_URL + parameters).text)
	
	
# @param sensor_id the id of the sensor e.g. ZBin1
# @param type the type of the sensor (WEIGHT_OBS_TYPE = 2 DIST_OBS_TYPE = 3)
# @param start_time the timestamp in which to start getting observations
# @param end_time the timestamp in which to stop getting observations
# @return an observation json object, sorted	
# http://sensoria.ics.uci.edu:8059/observation/get?sensor_id=ZBin1&type=2&start_timestamp=2018-07-01+00%3A00%3A00&end_timestamp=2018-07-25+00%3A00%3A00
# default timestamps are within the past 24 hours
def get_observation(sensor_id, type, start_time=None, end_time=None):
	if start_time is None:
		start_time = get_datetime(days=1)
	if end_time is None:
		end_time = getdatetime()
		
	parameters = urllib.parse.urlencode([("sensor_id", sensor_id), ("type", 2), ("start_timestamp", start_time), ("end_timestamp", end_time)])
	observations = eval(requests.get(session["tippershost"] + BASE_OBS_URL + parameters).text)
	return sorted(observations, key = lambda x : x['timestamp'])
	
# @param the id of the sensor e.g. ZBin1D
# @return the percentage fullness of the bin
def get_bin_fullness(bin_id):
	start = get_datetime(days = 1)
	end = get_datetime()
	obs = get_observation(bin_id, DIST_OBS_TYPE, start, end)
	if obs == []:
		return 0
	latest_ob = obs[0]
	for ob in obs:
		if ob["timestamp"] > latest_ob["timestamp"]:
			latest_ob = ob
	try:
		#fullness = (BIN_HEIGHT - latest_ob["payload"]["distance"]) / BIN_HEIGHT
		fullness = latest_ob["payload"]["distance"]
		if fullness < 0 or fullness > 1:
			fullness = 0
		return fullness
	except: 
		return 0
		
		
		
		
#helpers
def get_obs_type(sensor_type):
	if sensor_type == WEIGHT_SENSOR_TYPE:
		return WEIGHT_OBS_TYPE
	elif sensor_type == DIST_SENSOR_TYPE:
		return DIST_OBS_TYPE
	return -1